File: Solutions/3763-separate-squares-i/solution.py
class Solution:
    def separateSquares(self, squares: List[List[int]]) -> float:
        def area_below(y, squares):
            total = 0.0
            for xi, yi, li in squares:
                h = min(max(y - yi, 0), li)
                total += li * h
            return total
        total_area = sum(li * li for _, _, li in squares)
        target = total_area / 2.0
        lo = min(yi for _, yi, _ in squares)
        hi = max(yi + li for _, yi, li in squares)
        eps = 1e-5
        while hi - lo > eps:
            mid = (lo + hi) / 2.0
            if area_below(mid, squares) < target:
                lo = mid
            else:
                hi = mid
        return lo
        

File: Solutions/3763-separate-squares-i/solution.cpp
class Solution {
public:
    double epsilon = 1e-5;
    double check(vector<vector<int>>& squares, double vay) {
        double mass = 0;
        for(auto& sq: squares) {
            double x = sq[0], y = sq[1], len = sq[2];
            if(y >= vay) continue;
            if(y + len <= vay) {
                // complete zeer
                mass += (len*len);
            } else {
                // partial zeer
                mass += ((vay-y) * len);
            }
        }
        return mass;
    }

    double separateSquares(vector<vector<int>>& squares) {
        double total = 0;
        for(auto& sq: squares) total += ((double)sq[2] * sq[2]);
        double target = total / 2.0;
        double l = 0, r = 1e9;
        while(r - l > epsilon) {
            double mid = (l+r)/2.0;
            double ch = check(squares, mid);
            if(ch < target) l = mid;
            else r = mid;
        }
        return r;
    }
};
